### PR TITLE
feat: add GPT in markup mode

### PR DESCRIPTION
**Pull request by ReFFaT**

### CODE REVIEW

File: src/extensions/yfm/GPT/MarkupGpt/commands.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import type {EditorView} from '../../../../cm/view';
+
+import {HideMarkupGptExampleEffect, ShowMarkupGptExampleEffect} from './effects';
+
+export const showMarkupGptExample = (view: EditorView) => {
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/commands.ts:5]: remove "example" suffix or replace it with something else

File: src/extensions/yfm/GPT/MarkupGpt/commands.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    view.dispatch({effects: [ShowMarkupGptExampleEffect.of(null)]});
+};
+
+export const hideMarkupGptExample = (view: EditorView) => {
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/commands.ts:9]: remove "example" suffix or replace it with something else

File: src/extensions/yfm/GPT/MarkupGpt/effects.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import {StateEffect} from '../../../../cm/state';
+
+export const ShowMarkupGptExampleEffect = StateEffect.define();
+export const HideMarkupGptExampleEffect = StateEffect.define();
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/effects.ts:4]: remove "example" suffix or replace it with something else

File: src/extensions/yfm/GPT/MarkupGpt/plugin.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    ViewPlugin,
+    type ViewUpdate,
+} from '../../../../cm/view';
+import {ReactRendererFacet} from '../../../../markup';
+
+import {hideMarkupGptExample} from './commands';
+import {HideMarkupGptExampleEffect, ShowMarkupGptExampleEffect} from './effects';
+import {renderPopup} from './popup';
+
+const DECO_CLASS_NAME = 'react-popup-example-deco';
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/plugin.ts:19]: Specify relevant value

File: src/extensions/yfm/GPT/MarkupGpt/plugin.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            selectedPosition = {
+                from: 0,
+                to: 0,
+            };
+
+            constructor(view: EditorView) {
+                this._view = view;
+                this._renderItem = view.state
+                    .facet(ReactRendererFacet)
+                    .createItem('react-popup-example-in-markup-mode', () => this.renderPopup());
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/plugin.ts:60]: Specify relevant value for first argument

File: src/extensions/yfm/GPT/MarkupGpt/toolbar.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import {ToolbarDataType} from '../../../../toolbar';
+import {gptHotKeys} from '../constants';
+
+import {showMarkupGptExample} from './commands';
+
+export const mGptExampleToolbarItem: MToolbarSingleItemData = {
+    id: 'react-popup-example',
+    type: ToolbarDataType.SingleButton,
+    hotkey: gptHotKeys.openGptKeyTooltip,
+    title: 'Help with text',
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/toolbar.ts:12]: i18n?

File: src/extensions/yfm/GPT/MarkupGpt/toolbar.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import {MToolbarSingleItemData} from '../../../../bundle/config/markup';
+import {GPTIcon} from '../../../../icons';
+import {ToolbarDataType} from '../../../../toolbar';
+import {gptHotKeys} from '../constants';
+
+import {showMarkupGptExample} from './commands';
+
+export const mGptExampleToolbarItem: MToolbarSingleItemData = {
+    id: 'react-popup-example',
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/toolbar.ts:9]: Set relevant id

Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/toolbar.ts:9]: remove "example"

File: src/extensions/yfm/GPT/MarkupGpt/popup.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    onConfirmOk?: () => void;
+    onConfirmCancel?: () => void;
+} & GptDialogProps &
+    Pick<PopupProps, 'anchorRef' | 'container'>;
+
+export function renderPopup(anchor: HTMLElement, props: Props) {
+    const handleUpdate = (result?: CommonAnswer) => props.onUpdate?.(result);
+
+    return (
+        <Portal>
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/popup.tsx:21]: Unnecessary `<Portal>` wrapper

File: src/extensions/yfm/GPT/MarkupGpt/plugin.ts [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            }
+
+            update(update: ViewUpdate) {
+                if (update.docChanged || update.selectionSet) {
+                    this.decos = Decoration.none;
+                    return;
+                }
+
+                this.decos = this.decos.map(update.changes);
+                this.state = update.state;
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/plugin.ts:70]: It is unnecessary to save the entire state, it is enough to save only `from` and `to` from the selection

File: src/extensions/yfm/GPT/MarkupGpt/plugin.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+    toDOM() {
+        const spanElem = document.createElement('span');
+        spanElem.className = this.className;
+        spanElem.textContent = this.textContent;
+        return spanElem;
+    }
+}
+
+export function mGptExamplePlugin(gptProps: GptWidgetOptions) {
```
Comment (d3m1d0v) [inline src/extensions/yfm/GPT/MarkupGpt/plugin.ts:39]: "example"

**Review by d3m1d0v — CHANGES_REQUESTED** [review]

**Review by makhnatkin — APPROVED** [review]

File: src/extensions/additional/GPT/MarkupGpt/plugin.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            _onApplyResult(changedMarkup: string) {
+                const {from, to} = this.selectedPosition;
+                const changes = [{from: from, to: to, insert: changedMarkup}];
+
+                const transaction = this._view.state.update({
+                    changes: changes,
+                });
+
+                this._view.dispatch(transaction);
+                hideMarkupGpt(this._view);
```
Comment (d3m1d0v) [inline src/extensions/additional/GPT/MarkupGpt/plugin.ts:161]: You can add `{effects: [HideMarkupGptEffect.of(null)]}` to previous transaction and not call dispatch twice.

**Review by d3m1d0v — APPROVED** [review]

**Review by makhnatkin — APPROVED** [review]

**Pull request merged**